### ISSUE TITLE
Some preprocessing logic questions

### ISSUE THREAD

**Issue by chijames** [LayneIns/E2E-dialo-disentanglement#2]
1. In data_loader.py ln 37, why do we need to build_batch again (by setting add_noise=False)?
2. In utils.py ln 252, you are calculating new_labels using max_session_length. However, max_session_length here is calculated using noise_labels in ln 221, which might not be the case for padding new_labels. Is this the reason we need to build_batch again in Q1 (so we are using the correct max_session_length)?

Overall, if I understand correctly, then maybe we don't need the control logic (if add_noise: here and there) in build_batch function?

Thanks!

**Comment by chijames** [comment]
Also a minor observation: ln 265 to ln 272 can be merged into one return statement as they are doing the same thing.

**Comment by LayneIns** [comment]
@chijames 
Hi,

1. Because when we add noise to labels,  the noised labels are for generating the states of utterances. But for training, we have to provide the correct labels. The goal is that we want the model to predict the labels correctly even if the states are wrongly calculated.

2. For all repeated calculation related to "add_noise" in the code, the goal is to serve the above-mentioned purpose. In our paper we have a more detailed description for such a training method (known as Decision Sampling), of which the motivation is to bridge the discrepancy between training and inference.

3. Yes you are right. You can simply combine them as one.

**Comment by chijames** [comment]
Hi,

Thanks for the reply! Yeah I did read the decision sampling section in your paper. I think what confuses me is as stated in Q1. It seems to me that the only case we need to run ln 37 of data_loader.py is when max_session_length differs depending on add_noise=True / False. Otherwise, label_for_loss in ln34 and ln38 should be the same. If this is the case, then we don't need those "if add_noise" logic in build_batch since we are running the whole build_batch function again anyways. To be more concrete, what I am suggesting is like:

```
def build_batch(utterances, labels, word_dict, add_noise=False):
    max_conversation_length = max([len(x) for x in utterances]) # how many utterances in a session
    max_utterance_length = max([len(x) for one_uttr in utterances for x in one_uttr]) # max utterance length (how many words in a utterance)
    
    if add_noise:
        labels = add_noise_to_data(labels)
    max_session_length = max([max(Counter(l).values()) for l in labels])
    
    conversation_lengths = [len(x) for x in utterances]
    loss_mask = torch.arange(max_conversation_length).expand(len(conversation_lengths), max_conversation_length) \
                                < torch.LongTensor(conversation_lengths).unsqueeze(1)
    loss_mask = loss_mask.type(torch.int64)
    
    utterances_num, utterance_sequence_length = convert_utterances(utterances, word_dict)
    
    # session length, shape: [batch_size, max_labels]
    session_sequence_length = get_session_sequence_length(labels)
    
    # The state for a training case. Showing which of lstm state should be copied to the state matrix
    # [batch_size, max_conversation_length, max_labels+1]
    state_transition_matrix = build_state_transition_matrix(labels, max_conversation_length)

    # new_utterance_num_numpy: a numpy ndarray for the utterance number
    # new_labels: the label list after padding
    # new_utterance_sequence_length: utterance sequence length after padding 
    new_utterance_num_numpy, new_labels, new_utterance_sequence_length = \
                        padding_batch(utterances_num, labels, max_session_length, \
                                                    max_utterance_length, utterance_sequence_length)

    session_transpose_matrix = reorder_session(new_labels)

    label_for_loss = get_loss_labels(new_labels)

    return new_utterance_num_numpy, label_for_loss, new_labels, new_utterance_sequence_length, \
                session_transpose_matrix, state_transition_matrix, session_sequence_length, \
                    max_conversation_length, loss_mask
```

I should make it clear that this is not a bug, just want to make sure if I get the decision sampling logic right as the original logic is a bit confusing. Please correct me if I am wrong.

Thanks!

**Comment by LayneIns** [comment]
@chijames 
Hi, 

The code are necessary. If you check the value of "label_for_loss" in line 34 and line 38, they are actually different. This is why  we have to rerun line 38 if add_noise is set. The "label_for_loss" from line 34 is always useless.

**Comment by chijames** [comment]
Yeah, I totally agree with you that we need to get a new label_for_loss in line 38. But seems to me like we can simplify the logic in build_batch as all the "if add_noise:" have quite similar pattern. We just need one
```
if add_noise:
    labels = add_noise_to_data(labels)
```
in build_batch, then we call it twice outside in line 34 and 38. This makes the logic more clear. Anyway, I am closing this issue as it seems minor. Thanks!
